src/main/java/io/javalin/core/Plugin.kt

```diff
@@ -0,0 +1,25 @@
+package io.javalin.core
+
+import io.javalin.Javalin
+
+/**
+ * Add new functionality to javalin. Similar to [Extension].
+ * There are a few differences to extensions.
+ * 1. A Plugin needs to be registered during configuration time.
+ * 2. A Plugin will be initialized before all handlers. This allows to listen to all handler events
+ * that will occur in the application.
+ */
+@FunctionalInterface
+interface Plugin {
```

Comment:
Hi @TobiasWalle, sounds good to me!

Do you think it's possible to change the implememtation of `Extension` and make it behave like the plugin system?

I think for users, it will be less confusing to have just one API to do the same thing, though in two slightly different fashions.

Reply:
That is true. If you are okay with that, I would remove the `Extension` in favor of the Plugins and update the doc comment.

I thought about just extending the `Extension` but as the name `plugin` is used a lot in the javalin packages, I think to name this concept `Plugin` would be a better fit.

Reply:
the two differences between `Extension` and `Plugin` come due to the addition of `JavalinConfig`. When I wrote the extensions the, `JavalinConfig` did not exist yet. The ideas of both are very similar and I think it's reasonable to merge extensions and plugins into one.

I'd move `Extension` here and add `ExtensionLifecycleInit` - wdyt?

Reply:
Sounds good. But I think we should either
* rename the the package `io.javalin.plugin` to `io.javalin.extension`
* rename `Extension` to `Plugin`

I think otherwise it can be confusing to the user if the `OpenApiExtension` is in the `plugin` package, as both words have a very similar meaning.

@tipsy Do you have any thoughts on the naming?

Reply:
> rename `Extension` to `Plugin`

sounds good to me. I took over the names from [Sinatra extensions](http://sinatrarb.com/extensions.html#extending-the-dsl-class-context-with-sinatraregister). I'm ok with deprecating extensions.

How many people are out there using the current `Extension` APIs? Do you want to print deprecation warnings or remove them instantly?

Reply:
The current master is version 3.0, so we'll just remove them.

Reply:
Perfect. I will update the PR today or tomorrow.